module/src/main/scala/com/gu/googleauth/actions.scala

```diff
@@ -41,29 +48,139 @@ trait UserIdentifier {
+
+  /**
+    * Helper method that deals with sending a client for authentication. Typically this should store the target URL and
+    * redirect to the loginTarget. There shouldn't really be any need to override this.
+    */
+  def sendForAuth[A](request: RequestHeader) =
     Redirect(loginTarget).withSession {
-      request.session + (LOGIN_ORIGIN_KEY, request.uri)
+      request.session + (GoogleAuthFilters.LOGIN_ORIGIN_KEY, request.uri)
+    }
+
+  /**
+    * Redirects user to Google to start the login.
+    */
+  def startGoogleLogin()(implicit request: RequestHeader): Future[Result] = {
+    val antiForgeryToken = GoogleAuth.generateAntiForgeryToken()
+    GoogleAuth.redirectToGoogle(authConfig, antiForgeryToken).map {
+      _.withSession { request.session + (authConfig.antiForgeryKey -> antiForgeryToken) }
+    }
+  }
+
+  /**
+    * Extracts user from Google response and validates it, redirecting to `failureRedirectTarget` if the check fails.
+    */
+  def checkIdentity()(implicit request: RequestHeader): XorT[Future, Result, UserIdentity] = {
+    request.session.get(authConfig.antiForgeryKey) match {
+      case Some(token) =>
+        GoogleAuth.validatedUserIdentity(authConfig, token).attemptT.leftMap {
+          case e: IllegalArgumentException =>
+            Logger.warn("Login failure, anti-forgery token", e)
+            redirectWithError(failureRedirectTarget, "Login failure, anti-forgery token did not match", authConfig.antiForgeryKey, request.session)
+          case e: GoogleAuthException =>
+            Logger.warn("Login failure, GoogleAuthException", e)
+            redirectWithError(failureRedirectTarget, e.getMessage, authConfig.antiForgeryKey, request.session)
+          case e: Throwable =>
+            Logger.warn("Login failure", e)
+            redirectWithError(failureRedirectTarget, e.getMessage, authConfig.antiForgeryKey, request.session)
+        }
+      case None =>
+        Logger.warn("Login failure, anti-forgery token missing")
+        XorT.left[Future, Result, UserIdentity] {
+          Future.successful {
+            redirectWithError(failureRedirectTarget, "Login failure, anti-forgery token missing", authConfig.antiForgeryKey, request.session)
+          }
+        }
+    }
+  }
+
+  /**
+    * Looks up user's Google Groups and ensures they belong to any that are required. Redirects to
+    * `failureRedirectTarget` if the user is not a member of any required group.
+    */
+  def enforceGoogleGroups(userIdentity: UserIdentity, requiredGoogleGroups: Set[String], googleGroupChecker: GoogleGroupChecker)
+                         (implicit request: RequestHeader): XorT[Future, Result, Unit] = {
+    googleGroupChecker.retrieveGroupsFor(userIdentity.email).attemptT
+      .leftMap { t =>
+        Logger.warn("Login failure, Could not look up user's Google groups", t)
+        redirectWithError(failureRedirectTarget, "Unable to look up user's 2FA status", authConfig.antiForgeryKey, request.session)
```

Comment:
All these log messages and error messages related to 2FA should be rewritten to be more generic, as we are looking for an arbitrary set of required Google groups here.